### DISCUSSION TITLE
Change value of outputlist

### DISCUSSION THREAD

**Discussion by Je-Leo-AS** [Open-Agriculture/AgIsoStack-plus-plus#336] [Q&A] [answered]
Hello 
What's up?
I would like to change the index of an OutputString on the screen and would like to confirm if the function used for this is the bool VirtualTerminalClient::send_change_list_item(std::uint16_t objectID, std::uint8_t listIndex, std::uint16_t newObjectID). I tried to use this function but it didn't work. If so, could you explain what newObjectID is and if it could be a null value?

**Comment by ad3154** [comment]
Hello there!

It depends on what you mean by "address".

`send_change_list_item` Allows you to change what is shown in a list. The first parameter is the list you want to change, the second parameter is the index in the list to modify, and the third parameter is the new object that you want to place in that position in the list (or 0xFFFF, which is the null object ID if you want it empty). For example, in the [seeder example](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/tree/advanced-example/examples/seeder_example) I have an input list that contains 5 objects which looks like this:

![image](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/assets/10929341/c395fd8d-c3cd-4d7d-a957-1782c56885b8)

The options/output strings in the list are:

* Choose a statistic (which you can see in the image, since it's at index 0)
* CAN Bus
* Universal Terminal
* Task Controller
* Credits

So, in theory, if I wanted to remove "Universal Terminal", I could call `send_change_list_item` like this:
`send_change_list_item(statistics_InList, 2, UNDEFINED);` where `statistics_InList` is the object ID of the list, 2 is the index of "Universal terminal", and `UNDEFINED` is 0xFFFF.

Likewise, if I wanted to put some other output string where "Universal Terminal" is located, I'd call it like this:
`send_change_list_item(statistics_InList, 2, tcNumberSectionsSupported_OutStr);` where `statistics_InList` is the object ID of the list, `2` is the index of "Universal terminal", and `tcNumberSectionsSupported_OutStr` is the object ID of a different object.

Some limitations on this are defined by ISO11783-6:

- You can't add more things past the end of the list! So if the list contains 5 objects in the object pool, it can never contain more than that, so operations to change index 5+ will fail, so you'll want to consider that when designing the object pool.
- Indexes with 0xFFFF do not disappear, they are just not drawn by the VT, so if you set index 2 to 0xFFFF, that doesn't change the index of "Task Controller" in this example.
- Only some kinds of objects can be inside a list, which are defined in `Table A.2 — Allowed hierarchical relationships of objects`, so it's best to stick to output strings in my opinion.

If instead what you are trying to do is _move_ an output list somewhere else on the screen, you'll want either `send_change_child_location` or `send_change_child_position`

**Reply by Je-Leo-AS** [reply, marked as answer]
Got it!
So how do I change the index of the object shown in an OutputList? For example, I have an output list with 4 images and I need to change the index from 0 to 3 for each specific situation.

**Reply by alexandrekremer** [reply]
virtualTerminalClient->send_change_numeric_value(InputList_10000, 3);

Where:

send_change_numeric_value() is a VirtualTerminalClient method
InputList_10000 is #define InputList_10000                   10000
and 3 is the index to change

It works on my project where I have an InputList with ObjectID  = 10000

**Comment by ad3154** [comment]
> Got it!
> So how do I change the index of the object shown in an OutputList? For example, I have an output list with 4 images and I need to change the index from 0 to 3 for each specific situation.

If your output list has a valid number variable reference, then that's what controls the selected index, so you'd want to 'send_change_numeric_value' on that number variable. If you have the null object referenced in your output list, then the index is the value of the output list, and you'd want to 'send_change_numeric_value' on the output list's object ID.
